conn_browser.go

```diff
@@ -0,0 +1,292 @@
+// +build js,wasm
+
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errConnectionClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	errorHandler   *js.Func
+	mut            sync.Mutex
+	currDataMut    sync.RWMutex
+	currData       bytes.Buffer
+	closeOnce      sync.Once
+	closeSignal    chan struct{}
+	dataSignal     chan struct{}
+	localAddr      net.Addr
+	remoteAddr     net.Addr
+	firstErr       error
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:       raw,
+		closeSignal: make(chan struct{}),
+		dataSignal:  make(chan struct{}, 1),
+		localAddr:   NewAddr("0.0.0.0:0"),
+		remoteAddr:  getRemoteAddr(raw),
+	}
+	// Force the JavaScript WebSockets API to use the ArrayBuffer type for
+	// incoming messages instead of the Blob type. This is better for us because
+	// ArrayBuffer can be converted to []byte synchronously but Blob cannot.
+	conn.Set("binaryType", "arraybuffer")
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
+		return c.readAfterErr(b)
+	}
+
+	for {
+		c.currDataMut.RLock()
+		n, err := c.currData.Read(b)
+		c.currDataMut.RUnlock()
+		if err != nil && err != io.EOF {
+			// Return any unexpected errors immediately.
+			return n, err
+		} else if n == 0 || err == io.EOF {
```

Comment:
`n > 0 && err == io.EOF` is possible. In that case, we'd discard data.

This check should probably be:

```
if err != nil && err != io.EOF {
  return n, err // we read an error (is this even possible?)
}

if n != 0 {
  return n, nil // we read data, return it.
}

// wait for more data.
```

Reply:
Ah good point. Documentation for `bytes.Buffer.Read` makes it clear that the only possible error is `io.EOF`. I refactored the code based on your suggestion.